Chat Demo updated with Meeting Functionality

kmai-amazon opened the pull request:
**Issue #:**
We have provided AWS Chime customers with both a Chat demo and a Meeting demo. However, some customers are having difficulty combining the two into a single serverless client. So we've updated the Chat demo to include meetings functionality

**Description of changes:**
- Updated the apps/chat demo to support video and audio meetings
- Updated CF template in the apps/chat demo to deploy additional endpoints to the APIG to act as a meeting server with CreateMeeting, CreateAttendee, and EndMeeting functionality

**Testing**

1. How did you test these changes?
The changes were tested locally by following the steps in the updated README and in the blog post.
2. Can these changes be tested using one of the demo application? If yes, which demo application can be used to test it?
This change is a demo application

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

simmkyu reviewed (changes requested):
- Replace `Apache-2.0` to `MIT-0` in all files.
- Recommend moving the Lambda code from the CF template to individual files.
- PR includes a mix of JavaScript (js, jsx) and TypeScript (ts, tsx) files. I recommend using one language.

kmai-amazon commented on apps/chat/package.json:
```diff
@@ -4,7 +4,7 @@
   "scripts": {
     "lint": "eslint src --ext .js,.jsx --fix",
-    "start": "webpack serve"
+    "start": "npm install && webpack serve"
```
Removed

kmai-amazon commented on apps/chat/src/backend/serverless/template.yaml:
```diff
@@ -74,6 +74,95 @@ Resources:
+          ZipFile: >
+            "use strict";
+            const AWS = require("aws-sdk");
+            const uuidv4 = require("uuid");
```
Removed the import in ChimeAPI.js since it is not used

kmai-amazon commented on apps/chat/src/backend/serverless/template.yaml:
```diff
@@ -74,6 +74,95 @@ Resources:
+              console.log("Create Chime SDK App Admin");
+              if (event["RequestType"] === "Create") {
+                //create a chime app user
+                var createUserParams = {
```
Done

kmai-amazon commented on apps/chat/src/components/ActivityBar/index.tsx:
```diff
@@ -0,0 +1,30 @@
+// Copyright 2020-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
```
Done

kmai-amazon commented on apps/chat/src/backend/serverless/template.yaml:
```diff
@@ -697,6 +800,286 @@ Resources:
+            statusCode: 404,
+            headers: {
+              'Access-Control-Allow-Headers': 'Authorization',
+              'Access-Control-Allow-Origin': 'https://localhost:9000',
```
Yes, we are only allowing the localhost origin as that is what is specified in the demo instructions.

kmai-amazon commented on apps/chat/src/backend/serverless/template.yaml:
```diff
@@ -74,6 +74,95 @@ Resources:
+        Layers: 
+          - !Ref AWSSDKChimeLayer
+        Code:
+          ZipFile: >
```
The issue with moving the lambda source code to a separate file is that there is no automatic way (besides switching to CDK) to upload the lamdba sources to S3 for CF to use. Until we migrate this to CDK, by keeping all the lambda code in the CF template, it will always be up to date.

kmai-amazon commented:
Addressed all feedback. As for the mix of JavaScript and TypeScript, the original Chat demo was purely in JavaScript while the new code borrowed from the meetings demo is in TypeScript. Although ideally, this should all be in TypeScript, this is not a no-op task to convert all the JavaScript files.

kmai-amazon commented on apps/chat/src/backend/serverless/template.yaml:
```diff
@@ -74,6 +74,95 @@ Resources:
+                try {
+                  var admin_chime_response = await chime.createAppInstanceAdmin(
+                    createAdminParams,
+                    function (err, data) {
```
Updated

kmai-amazon commented on apps/chat/src/backend/serverless/template.yaml:
```diff
@@ -74,6 +74,95 @@ Resources:
+                }
+              } else {
+                //NOT A CREATE REQUEST
+                await response.send(event, context, response.SUCCESS, {});
```
Updated

kmai-amazon commented on apps/chat/src/backend/serverless/template.yaml:
```diff
@@ -697,6 +800,286 @@ Resources:
+        - !Ref AWSSDKChimeLayer
+      Runtime: nodejs12.x
+      Handler: index.handler
+      InlineCode: |
```
For the sake of the blog and keeping all the sources up-to-date, unless we switch to CDK, we would have to upload these to an S3 bucket and reference them in the CF. You're absolutely right that as separate files, we would be able to better run formatters and read the code but for the deployment, it would make keeping a consistent version for the CF templates and lamdbda handlers more difficult.

simmkyu commented:
> Addressed all feedback. As for the mix of JavaScript and TypeScript, the original Chat demo was purely in JavaScript while the new code borrowed from the meetings demo is in TypeScript.

Thanks for your comments. Other that that, PR looks good to me.

michhyun1 commented:
@kmai-amazon should this be merged?

kmai-amazon commented:
We are still waiting for approval to release alongside the blog.
